### ISSUE TITLE
Can't log into itch or GOG

### ISSUE THREAD

**Issue by fkrull** [flathub/net.lutris.Lutris#364]
Logging into itch or GOG doesn't work, as in, nothing happens when clicking the icon. Running from a terminal shows the following output:

```
$ flatpak run net.lutris.Lutris
2023-09-23 00:39:31,666: Starting Lutris 0.5.13
2023-09-23 00:39:31,668: Running AMD Mesa driver 23.1.6 on AMD Radeon Graphics (gfx1101, LLVM 15.0.7, DRM 3.54, 6.6.0-0.rc2.20230922gt27bbf45e.322.vanilla.fc39.x86_64) (0x747e)
2023-09-23 00:39:31,668: GPU: 1002:747E 1DA2:475D (amdgpu drivers)

** (lutris:2): WARNING **: 00:39:34.843: Failed to load shared library 'libwebkit2gtk-4.1.so.0' referenced by the typelib: /app/lib/libudev.so.1: version `LIBUDEV_247' not found (required by /usr/lib/x86_64-linux-gnu/libgudev-1.0.so.0)
Traceback (most recent call last):
  File "/app/lib/python3.10/site-packages/lutris/gui/widgets/sidebar.py", line 202, in on_connect_clicked
    self.service.login(parent=self.get_toplevel())
  File "/app/lib/python3.10/site-packages/lutris/services/base.py", line 359, in login
    dialog = WebConnectDialog(self, parent)
  File "/app/lib/python3.10/site-packages/lutris/gui/dialogs/webconnect_dialog.py", line 23, in __init__
    self.context = WebKit2.WebContext.new()
gi.repository.GLib.GError: g-invoke-error-quark: Could not locate webkit_web_context_new: 'webkit_web_context_new': /usr/lib/x86_64-linux-gnu/libjavascriptcoregtk-4.1.so.0: undefined symbol: webkit_web_context_new (1)
```

More info:
```
$ flatpak info net.lutris.Lutris

Lutris - Video game preservation platform

          ID: net.lutris.Lutris
         Ref: app/net.lutris.Lutris/x86_64/stable
        Arch: x86_64
      Branch: stable
     Version: 0.5.13
     License: GPL-3.0-or-later
      Origin: flathub
  Collection: org.flathub.Stable
Installation: system
   Installed: 981,4 MB
     Runtime: org.gnome.Platform/x86_64/44
         Sdk: org.gnome.Sdk/x86_64/44

      Commit: cee93d07f34a2245558bdf7425de84fc7540f047496bd01322c29ecb213d2fff
      Parent: c0dac5845e186ffa229e153f2f0426a19b769f6811df3ec5b9a77f8ec5f1a890
     Subject: Use org.freedesktop.ScreenSaver instead (ac483c2f)
        Date: 2023-07-31 00:34:15 +0000

$ flatpak info org.gnome.Platform//44

GNOME Application Platform version 44 - Shared libraries used by GNOME
applications

          ID: org.gnome.Platform
         Ref: runtime/org.gnome.Platform/x86_64/44
        Arch: x86_64
      Branch: 44
     License: GPL-2.0+
      Origin: flathub
  Collection: org.flathub.Stable
Installation: system
   Installed: 863,0 MB

      Commit: aaabdc8963d4f5dd7899676fa79e9a0e1664c0f7d96d23309eb2dff8eeceb694
      Parent: 584d7c3e03c7315975c6711d6503622c0b39c109b6faa89254a87d24cf918c6e
     Subject: Export org.gnome.Platform
        Date: 2023-09-21 00:41:12 +0000
```

**Comment by Willem-SN** [comment]
same

also mentioned on the forums:

https://forums.lutris.net/t/cant-connect-to-gog/18808

**Comment by Shaddox** [comment]
Good afternoon, 

For those looking for a migration to the debian package, all you have to do is just copy the contents of `~/.var/app/net.lutris.Lutris/data/lutris` into `~/.local/share/lutris` and from  `~/.var/app/net.lutris.Lutris/config` into `~/.config/lutris`  and stuff seems to work mostly fine, with the exception of the missing images and a few oddball games that seem they no longer want to work.

**Comment by strycore** [comment]
> Good afternoon,
> 
> For those looking for a migration to the debian package, all you have to do is just copy the contents of `~/.var/app/net.lutris.Lutris/data/lutris` into `~/.local/share/lutris` and from `~/.var/app/net.lutris.Lutris/config` into `~/.config/lutris` and stuff seems to work mostly fine, with the exception of the missing images and a few oddball games that seem they no longer want to work.

If you want the images, you also have to copy `~/.var/app/net.lutris.Lutris/data/lutris` to `~/.cache/lutris`.
For the games that no longer work, you may be missing dependencies that are in the Flatpak but not your host system.

**Comment by strycore** [comment]
I would appreciate help for the actual error causing this issue `/app/lib/libudev.so.1: version `LIBUDEV_247' not found`. 

I don't even know what version of udev is shipped in the Flatpak. Udev isn't referenced anywhere in the Flatpak definition.

**Comment by fkrull** [comment]
The soname is libudev.so.1.6.3, while the one in /usr/lib is libudev.so.1.7.4. There's another one with the same soname in /app/lib32, so my wild guess is it gets installed by a dependency that's included in both 32- and 64-bit versions. Which at least rules out some of the binaries?

**Comment by jficz** [comment]
not sure how this one works specifically, but shouldn't flatpacked apps be completely independent on system libs?

**Comment by fkrull** [comment]
> not sure how this one works specifically, but shouldn't flatpacked apps be completely independent on system libs?

They are (unless they go out of their way not to be). Flatpak apps use one of several shared runtimes so not every app has to build and bundle every library. The Lutris flatpak uses the GNOME runtime, for example. The libudev.so in /usr/lib that I was mentioning isn't from the host system but from the runtime.

> I don't even know what version of udev is shipped in the Flatpak. Udev isn't referenced anywhere in the Flatpak definition.

Found it, it's from the Wine base app. For some reason, that one includes libudev.so.1.6.3 (even though it's also in the runtime) and Lutris inherits that. It might be enough to just delete the file at build time. I'm trying that right now to see if it helps.

**Comment by D0han** [comment]
Just ran into that problem, last working version is  
Commit: 72ad7fdc1ad3cf6d58644cd02f40fbcdb7ee7f2c2315573a82eb3daeb35a7963
   Subject: Add 'setxkbmap' dependency (be0ae398)
      Date: 2022-11-26 22:33:45 +0000

After downgrading to it with `flatpak update --commit=72ad7fdc1ad3cf6d58644cd02f40fbcdb7ee7f2c2315573a82eb3daeb35a7963 net.lutris.Lutris` GOG works again as expected.

**Comment by AlmeidaWagner** [comment]
It's quite strange that the last working version is from November 2022. We have other issues open related to login issues as recent as August 2023, and I had set up my Amazon Games account in Lutris about 2 weeks ago just fine. I also did notice that libwebkit2gtk-4.1.so.0 is nowhere to be found inside the flatpak directory (/var/lib/flatpak/app/net.lutris.Lutris/current/active/files/lib/)

**Comment by jficz** [comment]
True. The original issue (and one I observed as well) was related to libwebkit, not libudev. How did this become related to libudev?

**Comment by AlmeidaWagner** [comment]
Could this be related to it using GNOME runtime platform version 44?

**Comment by D0han** [comment]
Possibly, as during the downgrade process also those were downgraded:
```
        ID                                                    Branch                   Op             Remote              Download
 1. [✓] org.freedesktop.Platform.GL.default                   21.08                    i              flathub             130.5 MB / 129.8 MB
 2. [✓] org.freedesktop.Platform.GL32.default                 21.08                    i              flathub             108.7 MB / 138.2 MB
 3. [✓] org.freedesktop.Platform.ffmpeg-full                  21.08                    i              flathub               4.7 MB / 4.2 MB
 4. [✓] org.freedesktop.Platform.ffmpeg_full.i386             21.08                    i              flathub               4.8 MB / 4.3 MB
 5. [✓] org.freedesktop.Platform.openh264                     2.0                      i              flathub               1.8 MB / 1.5 MB
 6. [✓] org.gnome.Platform.Compat.i386                        42                       i              flathub             166.5 MB / 158.4 MB
 7. [✓] org.gnome.Platform.Locale                             42                       i              flathub              18.0 kB / 336.9 MB
 8. [✓] org.gnome.Platform                                    42                       i              flathub             276.5 MB / 280.4 MB
 9. [✓] org.winehq.Wine.DLLs.dxvk                             stable-21.08             i              flathub              22.0 MB / 21.3 MB
10. [✓] org.winehq.Wine.gecko                                 stable-21.08             i              flathub             114.0 MB / 114.0 MB
11. [✓] org.winehq.Wine.mono                                  stable-21.08             i              flathub              47.2 MB / 85.1 MB
12. [✓] net.lutris.Lutris                                     stable                   u              flathub             241.1 MB / 301.6 MB
```

After i logged in to GOG and refreshed library, i had to upgrade Lutris to newest version to get game installed. Without upgrading it was consistently failing on wine and GLIBC_2.34

**Comment by fkrull** [comment]
> True. The original issue (and one I observed as well) was related to libwebkit, not libudev. How did this become related to libudev?

It became related to libudev because there's a warning clearly implicating libudev:

```
** (lutris:2): WARNING **: 00:39:34.843: Failed to load shared library 'libwebkit2gtk-4.1.so.0' referenced by the typelib: /app/lib/libudev.so.1: version `LIBUDEV_247' not found (required by /usr/lib/x86_64-linux-gnu/libgudev-1.0.so.0)
```

**Comment by fkrull** [comment]
> So it's essentialy GTK/libgudev packaging issue

No, libgudev expects to find the copy of libudev from the GNOME runtime. Instead it finds the copy of libudev that's included in this flatpak (inherited from the Wine flatpak). That one is older than libgudev expects and so it fails to load, which in turns makes webkitgtk not load.

**Comment by strycore** [comment]
> True. The original issue (and one I observed as well) was related to libwebkit, not libudev. How did this become related to libudev?

The stacktrace indicates libwebkit isn't able to load because of a version mismatch on udev:

```
** (lutris:2): WARNING **: 00:39:34.843: Failed to load shared library 'libwebkit2gtk-4.1.so.0' referenced by the typelib: /app/lib/libudev.so.1: version `LIBUDEV_247' not found (required by /usr/lib/x86_64-linux-gnu/libgudev-1.0.so.0)
...
gi.repository.GLib.GError: g-invoke-error-quark: Could not locate webkit_web_context_new: 'webkit_web_context_new': /usr/lib/x86_64-linux-gnu/libjavascriptcoregtk-4.1.so.0: undefined symbol: webkit_web_context_new (1)
```

**Comment by strycore** [comment]
If the udev mismatch comes from Wine then the fix is to update the Wine flatpak

**Comment by AlmeidaWagner** [comment]
I just checked the commit history for the wine version being used, and it last changed in the beginning of September, which would not explain why all these reports are from the last 10 days or so. Only GNOME Runtime changed in the meantime

**Comment by strycore** [comment]
Isn't the GNOME runtime not supposed to change between releases in order to not introduce API breakages like this one?
Maybe I can use @fkrull 's pull  request as a temporary workaround until Wine and the GNOME runtime we're using are in sync.

**Comment by AlmeidaWagner** [comment]
I would think so too. But this is the commit history for version 44, it looks quite extensive: https://gitlab.gnome.org/GNOME/gnome-build-meta/-/commits/gnome-44/

We can even see that they recently updated WebKitGTK. I wonder what would happen if lutris flatpak manifest specified an older "runtime-commit" to be used, instead of just GNOME 44

**Comment by fkrull** [comment]
I don't know why this problem has only surfaced now. Potentially one of the libraries in question (libgudev?) was updated and now relies on newer API features from libudev that it didn't need before. That would not have been a breaking change for consumers of the runtime.

But the source of the problem is quite clear, it's the bundled libudev in the Wine flatpak. Updating its version to match the version in the runtime would be a solution, but honestly I'm skeptical that Wine should even bundle libudev. They're not patching it or anything. This feels like a textbook case of "we added this dependency because it was necessary at the time but then didn't remove it when it became unnecessary". Which is why I'm so insistent on this, because keeping dependencies around that you don't know the reason for is a maintainability problem.

---
(I tried updating the Wine flatpak BTW, but unless I did it wrong, stable-23.08 still has the same too-old version of libudev.)

**Comment by strycore** [comment]
We install Wine to ensure that every dependency needed is available. Wine has a lot of 64bit and 32bit dependencies and the simplest way to have them all is to install Wine. If we were to remove Wine, we'd still need the dependencies.

For the same reason, Wine will need both 32 and 64bit versions of udev. The initial addition of udev was 3 years ago but the version  of udev itself was updated by @flathubbot 3 months ago. 

To add even more confusion, Wine uses eudev while the Freedesktop SDK uses libgudev.

**Comment by strycore** [comment]
If Wine's udev is removed then the conflict should be gone, doesn't matter which version of Freedesktop Lutris uses, as long as there's only one udev.
Can someone confirm if the Wine update fixes this?

**Comment by A6GibKm** [comment]
I am going to guess here, but udev was prob added in the wine extension because it needed a version newer than what was in available 22.08. I have a MR to use the GNOME 45 runtime at https://github.com/flathub/net.lutris.Lutris/pull/368, help would be welcome. Do note that GNOME 45 is based on top of 23.08.

**Comment by strycore** [comment]
I've tested the updated Flatpak and Webkit dialogs now show up properly
